### ISSUE TITLE
Commit 7455351 in dev build 4.2.6242 likely created issues with `__ptr32`

### ISSUE THREAD

**Issue by WeiN76LQh** [Vector35/binaryninja-api#6032]
**Version and Platform (required):**
- Binary Ninja Version: this is a problem on 4.2.6244-dev (c59961a8) and 4.2.6262-dev (5c271eff) but not on 4.2.6228-dev (312c3c99) and 4.1.5902 (f2165c5d)
- OS: macOS
- OS Version: 15.0
- CPU Architecture: M1

**Bug Description:**
It seems a pointer with the attribute `__ptr32` will initially consume 8 bytes of space in a structure and then shrinks down to 4 bytes. This is the case when creating a new structure from C source but not the case when creating a structure with the API `StructureBuilder`.

Creating a type with the following declaration:
```c
struct Foo {
	void *__ptr32 bar;
};
```

Will result in a type in the types view that looks like this:
```c
struct Foo
{
    void* __ptr32 bar;
                  ?? ?? ?? ??
};
```

`bar` is the correct size, at 4 bytes but `Foo` is 8 bytes in size with an extra 4 bytes of unused space after `bar`. Adding the `__packed` attribute to `Foo` does fix it *in this case* but this didn't use to be what happened. It used to be that `Foo` would be 4 bytes without the `__packed` attribute.

Creating a type with the following declaration:
```c
struct Foo {
	uint32_t bar;
	void *__ptr32 baz;
};
```

Will result in a type in the types view that looks like this:
```c
struct Foo
{
    uint32_t bar;
    __offset(0x4);
    void* __ptr32 baz;
};
```

`Foo` is the correct size at 8 bytes and `bar` and `baz` are at their correct offsets (`0` and `4` respectively). However there is this odd `__offset(0x4);` line in there.

What I'm finding in practice is that larger structures containing `__ptr32` pointers are sometimes having the next field after the pointer be deleted and replaced with `?? ?? ?? ??`. The field after the deleted field is at the correct offset, which is why I say it seems these types are starting out as consuming 8 bytes in size, resulting in them trashing the next field, but then shrinking down to consume 4 bytes of space.

**Comment by negasora** [comment]
Does setting the `__packed` attribute on the type fix the issue?

**Comment by WeiN76LQh** [comment]
> Does setting the `__packed` attribute on the type fix the issue?

Like I say, it does fix the issue in the simple case of:
```c
struct __packed Foo {
	void *__ptr32 bar;
};
```

But if you have`__ptr32` fields in the middle of a structure definition it causes problems for the field directly after.

Its probably possible to fix this issue by creating a separate packed structure for each instance of a `__ptr32` pointer, containing a single field for the pointer, and then using that structure for field types in other structures but thats a less than ideal solution for what appears to be a regression.

Like so:

```c
struct Foo __packed {
    void *__ptr32 bar;
};

struct Baz {
    Foo foobar;
};
```

**Comment by negasora** [comment]
Ah sorry, missed that

**Comment by CouleeApps** [comment]
I bet something in the way AdjustedType / AttributedType is handled has different alignment and therefore made the struct larger. Will look into this.

**Comment by CouleeApps** [comment]
Yep, looks like clang actually handled it fine, and the binja conversion code was missing a call to SetAlignment. Fixed in >=4.2.6269
